### PR TITLE
runtime: Configure the image request timeout to handle large workloads

### PR DESCRIPTION
**Pull request by ChengyuZhu6**
Efficiently deploying substantial workloads within confidential containers, encompassing multiple gigabytes, or even dozens of gigabytes of images, such as AI models, may present challenges regarding image pull timeouts. At present, the default timeout for request handling stands at 60 seconds. To customize this duration, it becomes necessary to adjust the parameter `defaultRequestTimeout` and subsequently recompile and replace the kata-shim with the compiled one.

The purpose of this Pull Request (PR) is to facilitate the configuration of image pull request timeout within the guest.

Fixes: https://github.com/kata-containers/kata-containers/issues/7389

Signed-off-by: ChengyuZhu6 <chengyu.zhu@intel.com>

### CODE REVIEW

File: src/runtime/virtcontainers/kata_agent.go
Comment (fidencio) [inline src/runtime/virtcontainers/kata_agent.go:368]: We want to override regardless if this is bigger or smaller than the default one.

**Comment by fidencio** [comment]
@ChengyuZhu6, patch looks very sensible to me and it's a good addition for sure.
Let's just make sure that if the user sets a value in the config, the value is respected regardless of being larger or smaller than the default one.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:368]: Thanks, @fidencio . I will update.

**Comment by ChengyuZhu6** [comment]
> patch looks very sensible to me and it's a good addition for sure.
> Let's just make sure that if the user sets a value in the config, the value is respected regardless of being larger or smaller than the default one.

Thanks, I agree with you.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @ChengyuZhu6!

**Review by dborquez — APPROVED** [review]

**Comment by ChengyuZhu6** [comment]
> @ChengyuZhu6, our "Commit Message Check" is failing as you need to have an issue opened for this, and then have a `Fixes: #${ISSUE_NUMBER}` as part of your commit message.
> 
> Please, adjust that and we can have this one merged ASAP.

Thanks, @fidencio . I have changed the commit message.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @ChengyuZhu6!

**Comment by ChengyuZhu6** [comment]
Hi , here is an error in `jenkins-ci-ubuntu-20.04_sev-x86_64-cc_sev_cri_containerd_k8s`:
 `14:06:10 fatal: could not create work tree dir '/home/jenkins/workspace/kata-containers-CCv0-ubuntu-20.04_sev-x86_64-CC_SEV_CRI_CONTAINERD_K8S-PR/go/src/github.com/kata-containers/tests': Permission denied`

Hi , here is an error in `jenkins-ci-ubuntu-20.04_snp-x86_64-cc_snp_cri_containerd_k8s `:
`4:18:10 ln: failed to create symbolic link '//opt/kata/share/defaults/kata-containers/configuration.toml': No such file or directory
14:18:10 [install_runtime.sh:112] ERROR: sudo ln -sf //opt/kata/share/defaults/kata-containers/configuration-qemu-snp.toml //opt/kata/share/defaults/kata-containers/configuration.toml`. I saw in the jenkins there has `./opt/confidential-containers/share/defaults/kata-containers/configuration.toml` , instead `//opt/kata/share/defaults/kata-containers/configuration.toml`

**Comment by fidencio** [comment]
> Here is an error in `jenkins-ci-ubuntu-20.04-x86_64-cc_cri_containerd `: `14:46:09 ln: failed to create symbolic link '//opt/kata/share/defaults/kata-containers/configuration.toml': No such file or directory 14:46:09 [install_runtime.sh:112] ERROR: sudo ln -sf //opt/kata/share/defaults/kata-containers/configuration-qemu.toml //opt/kata/share/defaults/kata-containers/configuration.toml 14:46:09 [install_kata.sh:75] ERROR: .ci/install_runtime.sh 14:46:09 [jenkins_job_build.sh:252] ERROR: ci/setup.sh`

We need to rebase this one atop of CCv0, as I just merged a PR there that will solve the issue, let me do that.

File: src/runtime/virtcontainers/kata_agent.go
Comment (c3d) [inline src/runtime/virtcontainers/kata_agent.go:369]: 1. The `if` is not needed with the current code, since you just copy the value.
2. Consider renaming `defaultRequestTimeout`, since it's no longer a default value.
3. Should we log something when this happens? Say someone entered a value of 1 second thinking it was 1 minute? Make it evident in the log that the timeout is short for a reason.

**Review by c3d — COMMENTED** [review]
This is a useful addition. However, I believe that the download timeout is highly image-dependent, therefore workload dependent, so you should consider adding an annotation to control it as well to make it possible to change the timeout on a per-workload basis.

**Comment by c3d** [comment]
@ChengyuZhu6 The code looks good to me as is, with a couple minor nits. So I could approve it for immediate merging if there is a hurry. On the other hand, if you have time, I think that taking one extra round to add annotation support would make this even more useful.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:369]: > 1. The `if` is not needed with the current code, since you just copy the value.

Thanks @c3d.  I'll remove the condition.

Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:369]: > 2. Consider renaming `defaultRequestTimeout`, since it's no longer a default value.

Maybe we can rename it to `RequestTimeout` or `ImageRequestTimeout`?

**Comment by ChengyuZhu6** [comment]
> @ChengyuZhu6 The code looks good to me as is, with a couple minor nits. So I could approve it for immediate merging if there is a hurry. On the other hand, if you have time, I think that taking one extra round to add annotation support would make this even more useful.

Thanks @c3d . I will implement

> @ChengyuZhu6 The code looks good to me as is, with a couple minor nits. So I could approve it for immediate merging if there is a hurry. On the other hand, if you have time, I think that taking one extra round to add annotation support would make this even more useful.

Thanks, @c3d. I'll fix the codes and add annotation support.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:369]: > 3. Should we log something when this happens? Say someone entered a value of 1 second thinking it was 1 minute? Make it evident in the log that the timeout is short for a reason.

 I agree to describe why the default timeout is short. But I haven't tested what the most appropriate timeout should be. Users should have the right to set a smaller value, if they want to get the error of the request as soon as possible.

Comment (c3d) [reply src/runtime/virtcontainers/kata_agent.go]: > > 3\. Consider renaming `defaultRequestTimeout`, since it's no longer a default value.
> 
> Maybe we can rename it to `RequestTimeout` or `grpcRequestTimeout`?

Usually, it's good if we have the same name as in the config. Makes searching the code easier. So I vote for either `ImageRequestTimeout`, or to change the config name as well. That being said, I don't see this as a 'grpc' timeout, since I think it covers just image requests, not a timeout on every single gprc request. Or does it?

Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go]: > > > 3. Consider renaming `defaultRequestTimeout`, since it's no longer a default value.
> > 
> > 
> > Maybe we can rename it to `RequestTimeout` or `grpcRequestTimeout`?
> 
> Usually, it's good if we have the same name as in the config. Makes searching the code easier. So I vote for either `ImageRequestTimeout`, or to change the config name as well. That being said, I don't see this as a 'grpc' timeout, since I think it covers just image requests, not a timeout on every single gprc request. Or does it?

`defaultRequestTimout` covers all grpc requests. So I add `imageRequestTimout` to cover image requests.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @ChengyuZhu6.

I'd like to hear from @c3d if all his comments have been addressed before merging.

**Comment by ChengyuZhu6** [comment]
> This is a useful addition. However, I believe that the download timeout is highly image-dependent, therefore workload dependent, so you should consider adding an annotation to control it as well to make it possible to change the timeout on a per-workload basis.

Ok. I'm working on add the image request time out to annotation

**Comment by ChengyuZhu6** [comment]
> @ChengyuZhu6 The code looks good to me as is, with a couple minor nits. So I could approve it for immediate merging if there is a hurry. On the other hand, if you have time, I think that taking one extra round to add annotation support would make this even more useful.

@c3d I have added annotation support to configure the image request timeout in the yaml.

File: src/runtime/virtcontainers/kata_agent.go
Comment (c3d) [inline src/runtime/virtcontainers/kata_agent.go:371]: Probably more an `Info` than a `Warn`. Unless there is some negative effect?

**Review by c3d — APPROVED** [review]
@ChengyuZhu6 Looks good. Please add a commit to update the list of annotations. https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-set-sandbox-config-kata.md.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:371]: Ok. I'll update it.

**Comment by ChengyuZhu6** [comment]
> @ChengyuZhu6 Looks good. Please add a commit to update the list of annotations. https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-set-sandbox-config-kata.md.

Thanks, I'll add update the doc.

**Comment by fidencio** [comment]
@c3d, thanks for the quick review on this one.

I will re-run the tests and get it merged, thanks @ChengyuZhu6!

File: src/runtime/virtcontainers/kata_agent.go
Comment (amshinde) [inline src/runtime/virtcontainers/kata_agent.go:371]: Can you merge this in the previous commit?

Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go:371]: Ok.

**Comment by fidencio** [comment]
@UnmeshDeodhar, http://jenkins.katacontainers.io/job/kata-containers-CCv0-ubuntu-20.04_sev-x86_64-CC_SEV_CRI_CONTAINERD_K8S-PR/877/consoleFull has passed it seems it failed due to the artefacts not being cleaned up.

Please, let me know if we can merge it.

**Comment by UnmeshDeodhar** [comment]
@fidencio the console output has this:
`01:48:33 INFO: Run tests
01:48:33 1..5
01:51:09 ok 1 [cc][kubernetes][containerd][sev] Test SEV unencrypted container launch success
01:51:31 ok 2 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch failure with INVALID measurement
01:51:36 ok 3 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch success with NO measurement
01:51:38 ok 4 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch success with VALID measurement
01:51:49 ok 5 [cc][kubernetes][containerd][sev] Test SEV-ES encrypted container launch success with VALID measurement
`
So, looks like the tests are passing. You can merge this.

**Pull request merged**
